source/common/router/scoped_config_impl.cc

```diff
@@ -3,6 +3,107 @@
 namespace Envoy {
 namespace Router {
 
+bool ScopeKey::operator!=(const ScopeKey& other) const { return !(*this == other); }
+
+bool ScopeKey::operator==(const ScopeKey& other) const {
+  if (this->fragments_.empty() || other.fragments_.empty()) {
+    // An empty key equals to nothing, "NULL" != "NULL".
+    return false;
+  }
+  return std::equal(this->fragments_.begin(), this->fragments_.end(), other.fragments_.begin(),
+                    other.fragments_.end(),
+                    [](const std::unique_ptr<ScopeKeyFragmentBase>& left,
+                       const std::unique_ptr<ScopeKeyFragmentBase>& right) -> bool {
+                      // Both should be non-NULL now.
+                      return *left == *right;
+                    });
+}
+
+HeaderValueExtractorImpl::HeaderValueExtractorImpl(
+    ScopedRoutes::ScopeKeyBuilder::FragmentBuilder config)
```

Comment:
Nit: style guide says move semantics in an arg imply `&&` at the end of the type, see https://github.com/envoyproxy/envoy/blob/master/STYLE.md.

Reply:
done